src/client.rs

```diff
@@ -0,0 +1,162 @@
+use cosmian_crypto_core::{
+    reexport::rand_core::SeedableRng, Aes256Gcm, CryptoCoreError, CsRng, Dem,
+    FixedSizeCBytes, Instantiable, Nonce, RandomFixedSizeCBytes, SymmetricKey,
+};
+use rand::{Rng, RngCore};
+use std::io::{Error, ErrorKind};
+
+use crate::{btree::DataItem, oram::BUCKET_SIZE};
+
+pub struct ClientORAM {
+    pub stash: Vec<DataItem>,
+    csprng: CsRng,
+    key: SymmetricKey<{ Aes256Gcm::KEY_LENGTH }>,
+}
+
+impl ClientORAM {
+    pub fn new() -> ClientORAM {
+        let mut csprng = CsRng::from_entropy();
+        let key = SymmetricKey::new(&mut csprng);
+
+        ClientORAM {
+            // Empty stash at initialization.
+            stash: Vec::new(),
+            csprng,
+            key,
+        }
+    }
+
+    pub fn gen_key(&mut self) {
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
+
+    pub fn generate_dummies(
+        &mut self,
+        nb_dummies: usize,
+        block_size: usize,
+    ) -> Result<Vec<DataItem>, Error> {
+        if !(nb_dummies + 1).is_power_of_two() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Number of blocks shall be power of 2 minus one".to_string(),
```

Comment:
I see your point, since the client should be able to generate dummies **and** insert encrypted data, it makes sense not to restrict the number of dummy items to generate.

Done in f96fcb0d813824aa61c447b1d01f2bdef548892b